api/graphql/reservations/reservation_serializers.py

```diff
@@ -402,6 +425,92 @@ def check_reservation_kind(self, reservation_unit):
                 "since its reservation kind is SEASON."
             )
 
+    def requires_price_calculation(self, data):
+        # If pk is not given, this is a create request -> price is always calculated
+        if "pk" not in data:
+            return True
+
+        if "begin" in data and self.instance.begin != data["begin"]:
+            return True
+
+        if "end" in data and self.instance.end != data["end"]:
+            return True
+
+        if "reservation_unit" in data:
+            existing_unit_ids = []
+            for unit in self.instance.reservation_unit.all():
+                existing_unit_ids.append(unit.pk)
+
+            new_unit_ids = []
+            for unit in data["reservation_unit"]:
+                new_unit_ids.append(unit.pk)
+
+            if set(existing_unit_ids) != set(new_unit_ids):
+                return True
+
+        return False
+
+    def calculate_price(
+        self,
+        begin: datetime.datetime,
+        end: datetime.datetime,
+        reservation_units: List[ReservationUnit],
+    ) -> PriceCalculationResult:
+        price_unit_to_minutes = {
+            ReservationUnit.PRICE_UNIT_PER_15_MINS: 15,
+            ReservationUnit.PRICE_UNIT_PER_30_MINS: 30,
+            ReservationUnit.PRICE_UNIT_PER_HOUR: 60,
+            ReservationUnit.PRICE_UNIT_PER_HALF_DAY: 720,
+            ReservationUnit.PRICE_UNIT_PER_DAY: 1440,
+            ReservationUnit.PRICE_UNIT_PER_WEEK: 10080,
+        }
+
+        total_reservation_price: Decimal = Decimal(0.0)
+
+        first_paid_unit_price: Decimal = 0.0
+        first_paid_unit_tax_percentage: Decimal = 0.0
+        is_first_paid_set = False
+
+        for reservation_unit in reservation_units:
+            # If unit is not paid, there is no need for calculations. Skip.
```

Comment:
This wording can confuse: "If unit is not paid,". I suggest changing it to "If unit pricing type is not PAID,"